doc/source/advanced/options.rst
Options
=======

MALT supports several options which can be passed by **file** or **command line**.

The options are represented into a **two level** tree with **groups** and **options**
to follow the **ini** file standard.

Command line
------------

The **command line** options follow the simple semantic :

.. code-block:: shell

    # pass the options at launch time
    malt -o {GROUPE}:{OPTION}={VALUE} -o {GROUPE2}:{OPTION2}={VALUE2} ./my_program

    # In a single option
    malt -o "{GROUPE}:{OPTION}={VALUE};{GROUPE2}:{OPTION2}={VALUE2}" ./my_program

Configuration file
------------------

You can provide a **config file** to MALT to setup some features. This file uses the **INI**
format. With the malt script :

.. code-block:: shell

    malt -c config.ini {YOUR_PROGRAM} [OPTIONS]

Note you can use **file** plus override some options by adding `-o/--option` afterward :

.. code-block:: shell

    malt -c config.ini -o time:points=5000 {YOUR_PROGRAM} [OPTIONS]

Available options
-----------------

Example of config file :

.. code-block:: ini

    [time]
    enabled=true           ; enable time profiles
    points=512             ; keep 512 points
    linar-index=false      ; use action ID instead of time

    [stack]
    enabled=true           ; enable stack profiles
    mode=backtrace         ; select stack tracing mode (backtrace|enter-exit)
    resolve=true           ; Automatically resolve symbols with addr2line at exit.
    libunwind=false        ; Enable of disable usage of libunwind to backtrace.
    skip=4                 ; Number of stack frame to skip in order to cut at malloc level

    [sampling]
    enabled=false          ; Sample and instrument only some stack.
    bandwidth=4093         ; Instrument the stack when seen passed 4K-3 bytes of alloc requests (ideally should be prime number).
    count=571              ; Instrument the stack when seen passed 4K-3 bytes of alloc requests (ideally should be prime number).

    [addr2line]
    bucket=350             ; Handle the addr2line calls by buckets and treat each bucket in parallel.
    threads=8              ; Number of threasd to use to call addr2line in parallel.
    huge=52428800          ; For larger elf files, do not treat them in parallel nor buckets.

    [output]
    name=malt-%1-%2.%3     ; base name for output, %1 = exe, %2 = PID, %3 = extension
    lua=true               ; enable LUA output
    json=true              ; enable json output
    callgrind=true         ; enable callgrind output
    indent=false           ; indent the output profile files
    config=true            ; dump current config
    verbosity=default      ; malt verbosity level (silent, default, verbose)
    stack-tree=false       ; store the call tree as a tree (smaller file, but need conversion)
    loop-suppress=true     ; Simplify recursive loop calls to get smaller profile file if too big
    verbosity=default      ; Malt verbosity level (silent, default, verbose).
	stack-tree=false       ; Store the call tree as a tree (smaller file, but need conversion in the reader);

    [max-stack]
    enabled=true           ; enable of disable strack size tracking (require -finstrument-functions)

    [distr]
    alloc-size=true        ; generate distribution of allocation size
    realloc-jump=true      ; generate distribution of realloc jumps

    [trace]
    enable=false           ; enable dumping allocation event tracing (not yet used by GUI)

    [info]
    hidden=false           ; try to hide possible sensible names from profile (exe, hostname...)

    [filter]
    exe=                   ; Only apply malt on given exe (empty for all)
    childs=true            ; Instrument child processes or not
    enabled=true           ; Enable or disable MALT when threads start
    ranks=                 ; Instrument only the given ranks from list as : 1,2-4,6

    [dump]
    on-signal=             ; Dump on signal. Can be comma separated list from SIGINT, SIGUSR1,
                           ; SIGUSR2... help, avail (limited to only one dump)
    after-seconds=0        ; Dump after X seconds (limited to only one time)
    on-sys-full-at=        ; Dump when system memory become full at x%, xG, xM, xK, x  (empty to disable).
    on-app-using-rss=      ; Dump when RSS of the app reach the given limit in %, G, M, K (empty to disable).
    on-app-using-virt=     ; Dump when Virtual Memory of the app reach limit in %, G, M, K (empty to disable).
    on-app-using-req=      ; Dump when Requested Memory of the app reach limit in %, G, M, K (empty to disable).
    on-thread-stack-using= ; Dump when one stack reach limit in %, G, M, K (empty to disable).
    on-alloc-count=        ; Dump when number of allocations reach limit in G, M, K (empty to disable).
    watch-dog=false        ; Run an active thread spying continuouly the memory of the app, not only sometimes.

    [python]
    enabled=true           ; Enable of disable python instrumentation.
    stack=enter-exit       ; Select the Python stack instrumentation mode (backtrace, enter-exit, none).
    mix=false              ; Mix C stack with the python ones to get a uniq tree instread of two distincts
                           ; (not this adds overhead).
    obj=true               ; Instrument of not the OBJECT allocator domain of python.
    mem=true               ; Instrument of not the MEM allocator domain of python.
    raw=true               ; Instrument of not the RAW allocator domain of python.
    hide-imports=true      ; Hide the stack of memory allocations done by imports.
	mode=profile           ; Define the python instrumentation mode which define how to get the line number ('profile' or 'trace').

    [c]
    malloc=true            ; Track C malloc.
    mmap=true              ; Track C mmap direct calls.

    [tools]
    nm=true                ; Enable usage of NM to find the source locatoin of the global variables.
    nm-max-size=50M        ; Do not call nm on .so larger than 50 MB to limit the profile dump overhead.

Section `time`
--------------

This set of option permits to configure the time charts about the dynamic of the application.

Option `time:enabled`
^^^^^^^^^^^^^^^^^^^^^

Enable support of tracking state values over time to build time charts.

**Default**: true.

.. code-block:: shell

    malt -o time:enabled=true ./my_program
    malt -o time:enabled=false ./my_program

Option `time:points`
^^^^^^^^^^^^^^^^^^^^

Define the number of points used to discretized the execution time of the application.

**Default**: 512.

.. code-block:: shell

    malt -o time:points=512 ./my_program

Option `time:linear`
^^^^^^^^^^^^^^^^^^^^

Do not use time to index data but a linear value increased on each call (might be interesting
not to shrink intensive allocation steps on a long program which mostly not do allocation over the run.

**Default**: false.

.. code-block:: shell

    malt -o time:linear=true ./my_program
    malt -o time:linear=false ./my_program

Section `stack`
---------------

Option `stack:enabled`
^^^^^^^^^^^^^^^^^^^^^^

Enable support of stack tracking.

**Default**: true.

.. code-block:: shell

    malt -o stack:enabled=true ./my_program
    malt -o stack:enabled=false ./my_program

Option `stack:mode`
^^^^^^^^^^^^^^^^^^^

Override by `-s` option from the command line, it set the stack tracking method to use. See `-s` documentation
for more details. The available values are `enter-exit`, `backtrace`, `python`, `none`. By default backtrace
is used as it works out of the box everywhere. It is slower than enter-exit but this last one needs recompilation
of the code with `-finstrument-functions` and see only the libs recompiled with this option. `python` permit
to project all the C allocation directly into the python stack domain so it hides the C underwood. Good to
improve performance over python when you are not interested into the C details.

**Default**: backtrace.

.. code-block:: shell

    malt -o stack:mode=backtrace ./my_program
    malt -o stack:mode=enter-exit ./my_program
    malt -o stack:mode=python ./my_program
    malt -o stack:mode=none ./my_program

Option `stack:resolve`
^^^^^^^^^^^^^^^^^^^^^^

Enable symbol resolution at the end of execution to extract full names and source location if debug options
is available.

**Default**: true.

.. code-block:: shell

    malt -o stack:resolve=true ./my_program
    malt -o stack:resolve=false ./my_program

Option `stack:libunwind`
^^^^^^^^^^^^^^^^^^^^^^^^

Use linunwind backtrace method instead of the one from glibc.

**Default**: false.

.. code-block:: shell

    malt -o stack:libunwind=true ./my_program
    malt -o stack:libunwind=false ./my_program

Option `stack:skip`
^^^^^^^^^^^^^^^^^^^

In backtrace mode, the backtrace is called inside our instrumentation function itself called by malloc/free....
In order to skip our internal code we need to remove them. But on some compilers or distros, inlining and LTO
configuration make wrong detection. This can be fixed by playing with this value to keep malloc/free/calloc
as leaf for every calltree extracted by MALT.

If you don't see malloc as leaf, you should increase this value. Decrease if you start to see MALT internal
function inside.

Notice in some cases (Gentoo, Gcc-8) we start to see LTO trashing totaly the malloc/free call in O3 mode 
normaly seen via backtrace. There is currently no other solution except recompiling in O0/O1 to avoid
or maybe disable LTO optimizations or consider not having the exact location of the malloc itself.

**Default**: 4.

.. code-block:: shell

    malt -o stack:skip=4 ./my_program

Option `sampling:enabled`
^^^^^^^^^^^^^^^^^^^^^^^^^

Enable sampling mode for the stacks by captuing only for a few mallocs. It is less accurate than a full
profile but cost less in memory and CPU. In sampling mode, the stack is processed only when we seen passed
a few bytes allocated, otherwise we consider the last seen call stack. With python you should also
enable the backtrace mode for solving stacks via `python:stack=backtrace`.

It will use the options `sampling:volume` and `sampling:count` to know at which rate to sample.

**Default**: false.

.. code-block:: shell

    malt -o sampling:enabled=true ./my_program
    malt -o sampling:enabled=false ./my_program

Option `sampling:volume`
^^^^^^^^^^^^^^^^^^^^^^^^

Define the amount of data seen passed between two samples. Idealy this should be a prime number to avoid
some multiple base biases..

It is completed by also sampling on count with `sampling:count`.

**Default**: 4093.

.. code-block:: shell

    malt -o sampling:volume=4093 ./my_program
    malt -o sampling:volume=5242883 ./my_program
    malt -o sampling:volume=10485767 ./my_program
    malt -o sampling:volume=20971529 ./my_program

Option `sampling:count`
^^^^^^^^^^^^^^^^^^^^^^^

Define the number of operations passing between two sampling. Idealy this should be a prime number to avoid
some multiple base biases..

It is completed by also sampling on count with `sampling:volume`.

**Default**: 571.

.. code-block:: shell

    malt -o sampling:count=13 ./my_program
    malt -o sampling:count=31 ./my_program
    malt -o sampling:count=67 ./my_program
    malt -o sampling:count=67 ./my_program
    malt -o sampling:count=257 ./my_program
    malt -o sampling:count=571 ./my_program

Option `addr2line:bucket`
^^^^^^^^^^^^^^^^^^^^^^^^^

Define the number of addresses to place in the bucket to resolve while calling `addr2line`.
Each bucket is executed in parallel so it goes faster. Except for the **huge** ELF files which
are treated as a single bucket and in sequentiel, not to kill the memory in MPI and not
loading the large file many times.

**Default**: 350.

.. code-block:: shell

    malt -o addr2line:bucket=350 ./my_program

Option `addr2line:threads`
^^^^^^^^^^^^^^^^^^^^^^^^^^

Define the number of threads to use to make multiple calls to `addr2line` in parallel.

**Default**: 8.

.. code-block:: shell

    malt -o addr2line:threads=8 ./my_program

Option `addr2line:huge`
^^^^^^^^^^^^^^^^^^^^^^^

Define the threashold for bucket splitting. For files larger than this limit, consider them
as huge and not using nor buckets, not parallelism not to trash the memory of the node with
too many large memory usage processes (especially in MPI).

**Default**: 52428800 (50 MB).

.. code-block:: shell

    malt -o addr2line:huge=52428800 ./my_program

Section `output`
----------------

Option `output:name`
^^^^^^^^^^^^^^^^^^^^

Define the name of the profile file. %1 is replaced by the program name, %2 by the PID or MPI rank and %3 by extension.


ChangeLog.md
Change log
==========

1.6.3 - 18/08/2026
------------------

 - Increment some Javascript, Rust and Python deps versions for security
 - Fix documentation issues.
 - Fix labels on webview charts
 - Fix typos in webview
 - Validate on Ubuntu-26.04

1.6.2 - 14/06/2026
------------------

 - Update JavaScript NPM libraries
 - Fix Dummy Mutex building
 - Work on tracing API in the backend part.
 - Fix stack:mode being always C stack and never Python as it should.
 - Make compiling with gcc-7.x supported again
 - Fix portability to old SUSE/leap-15.6 and add it to the checklist.
 - Fix relative path in the GUI to be used inside code_server with port redirection.

1.6.1 - 08/02/2026
------------------

 - Fix possible gliches on physical and virtual memory measurements.
 - Minor fix on ont time chart.
 - Install the examples.
 - Fix a buffer overflow due to forgetting addint \0 after calling readlink.
 - Display profile file name in the home page of the webview.
 - Add a warning when using metric local.peak + inclusive which is not an exact computation.

1.6.0 - 29/01/2026
------------------

 - Do not tread in parallel the huge ELF files with addr2line (-o addr2line:huge=50M).
 - With addr2line tread in a single shot the huge ELF files.
 - With addr2line pass via a file to transmit the addresses for huge ELF files.
 - Fix portability on RedHat 8,9 and 10.
 - Improve documentation.
 - Refactor the option management class and rename some options to get consistency in names.
 - Fix broken reader public API (not yet stable).
 - Add possibility to dump the webview in a static mode (via --static & --static-summary).

1.5.0 - 07/01/2025
------------------

 - Replace the old webview in Angular 2.0 by a new one in VueJS.
 - Many fixes about python support, now should work relatively correctly.
 - Write the documentation in sphinx format.
 - Remove usage of login/password by a token
 - Remove malt-passwd command.
 - Remove dependency to libcrypto.
 - Rename maltDisable/Enable => malt_enable/malt_disable.
 - Add option to use embeded options or system onces.
 - Use iniparser of the system.
 - Make patches to get in better state for a debian packaging close to offical way.
 - Hide internal symbols with -fvisbility
 - Add build option --disable-webview and --disable-profiler

1.4.1 - 14/10/2025
------------------

 - Fix blocking issue in malt-passwd command (issue #136).

1.4.0 - 26/09/2025
------------------

 - Enable python support by default if python lib and headers are found.
 - Replace the webiew backend in NodeJS by a C++ one (faster and no deps to NodeJS required).
 - Add MPI rank filtering via -o filter:ranks=1,2-4,6 to only instrument some MPI ranks.
 - Add instrumentation of mmap / munmap / mremap.
 - Add more exemples.
 - Report python & C domains counters in profile.
 - Hide calls to PyDataMem_\*New\* in reports.
 - Remove malt-qt5.
 - Fix bugs on chart axis rendering in webview.
 - Fix some build issues when tests are disabled.
 - Fix broken thread tracking such that threads was not seen in the report.
 - Fix some issues in some page rendering (annotations.)
 - Fix segfault in calling addr2line in parallel.
 - Fix bugs about python support.

1.3.1 - 20/06/2025
------------------

 - Add python:hideImports option to group the imports in a single target not to look them (enabled by default).
 - Document tools:nm & tools:nmMaxSize.
 - By default disable nm usage on profile python-only & python-default.
 - Fix mistake which make the embeded google test installed with MALT.

1.3.0 - 16/06/2025
------------------

 - Add python profiling support (experimental, disabled by default, need to --enable-python at build time).
 - Check portability on Ubuntu (22.04, 22.10, 24.04, 25.04), Debian (10, 11, 12), Fedora (40, 41), Arch, Gentoo
 - Add more unit tests.
 - Refactor the injection and init of MALT in a proper way with guards.
 - Reshape the wrapper / inject / state code and files in a more ordered way.
 - Rewrite the symbol solver infra code in a proper way.
 - Add dumping options (on-sys-full-at, on-app-using-rss, on-app-using-virt, on-app-using-req, on-thread-stack-using).
 - Add parameter "tools:nm" and "tools:nmMaxSize" to avoid analysing large files of PyTorch by default
   to get source location of global variables.. 
 - Upgrade cmake dependency to 3.5 to avoid warning a recent systems.
 - Rename license file to be more standard.
 - Introduce usage of JeMalloc as the internal allocator in place of the custom one (required for large python apps).
 - Use LTO & fvisibility=hidden in release mode when tests are disabled (default).
 - Optimize the stack tree browser in the source page to handle large packages.
 - Add --gdb option on malt command to ease debugging.
 - Add progress bar for profile loading in webview.
 - Report the % of physical mapping of global variables.
 - Better terminal output for the webview and display + handle URL only when profile is loaded, not when not ready.
 - Parallel solving of symbols by calling several addr2line at a time.
 - Fix NodeJs limitation to load large JSON files by increasing its allowed memory limit.
 - Fix an issue with TLS at init which induce an infinite loop in some conditions.
 - Fix some issues in the JSON file generated.
 - Fix in the webview, size in base 2 instead of 10.
 - Fix in the webview, NaN over global variables and TLS.
 - Fix TLS size not shown correctly in summary.
 - Fix some code warnings.
 - Fix some missing units on charts in webview.

1.2.6 - 14/03/2025
------------------

 - Fix issue requiring NPM when not required if not wanting the webview.

1.2.5 - 05/02/2025
------------------

 - Fix broken interface from 1.2.4.
 - Introduce some extra unit-tests.

1.2.4 - 03/10/2024
------------------

 - Fix issue in previous release making some unit tests broken (line numbers) by file header changing.
 - Fix cmake issue with libunwind find package under Gentoo with x86_64 + x86.
 - Add build summary in CMake script.


src/lib/common/Options.cpp
/***********************************************************
*    PROJECT  : MALT (MALoc Tracker)
*    VERSION  : 1.2.6
*    DATE     : 09/2024
*    LICENSE  : CeCILL-C
*    FILE     : src/lib/common/Options.cpp
*-----------------------------------------------------------
*    AUTHOR   : Sébastien Valat - 2014 - 2024
*    AUTHOR   : Sébastien Valat (ECR) - 2014 - 2015
*    AUTHOR   : Sébastien Valat (CERN) - 2015
***********************************************************/

/**********************************************************/
//std
#include <cstdio>
#include <cassert>
//internals
#include "Debug.hpp"
#include "Options.hpp"
#include <json/JsonState.h>

/**********************************************************/
namespace MALT 
{

/**********************************************************/
Options * gblOptions = NULL;
/**
 * To be used in LocalAllocStackProfiler to cut MALT internal function calls
 * when backtracing and stop exactly at malloc/free level.
 * Made as an option as I already had serveral time issues depending on compiler
 * inlining operations so in case it append to someone he can fix without
 * recompiling.
**/
static const int cstDefaultStackSkip = 4;

/**********************************************************/
/**
 * To convert MALT verbosity level.
**/
static const char * cstVerbosityLevels[MALT_VERBOSITY_COUNT] = {
	"silent",
	"default",
	"verbose"
};

/**********************************************************/
/**
 * Constructor to setup the default values for each options
**/
Options::Options(void)
{
	//stack
	this->stackProfileEnabled     = true;
	this->stackResolve            = true;
	this->stackLibunwind          = false;
	this->stackMode               = "backtrace";
	this->stackSkip               = cstDefaultStackSkip;
	//time
	this->timeProfileEnabled      = true;
	this->timeProfilePoints       = 512;
	this->timeProfileLinear       = false;
	//output
	this->outputName              = "malt-%1-%2.%3";
	this->outputIndent            = false;
	this->outputJson              = true;
	this->outputLua               = false;
	this->outputCallgrind         = false;
	this->outputDumpConfig        = false;
	this->outputVerbosity         = MALT_VERBOSITY_DEFAULT;
	this->outputStackTree         = false;
	this->outputLoopSuppress      = false;
	//max stack
	this->maxStackEnabled         = true;
	//maps
	this->distrAllocSize          = true;
	this->distrReallocJump        = true;
	//trace
	this->traceEnabled            = false;
	//info
	this->infoHidden              = false;
	//filter
	this->exe                     = "";
	this->childs                  = true;
	this->enabled                 = true;
	//dump
	this->dumpOnSignal            = MALT_NO_DUMP_SIGNAL;
	this->dumpAfterSeconds        = 0;
}

/**********************************************************/
/**
 * Manage operator == to help validation in unit test suite.
**/
bool Options::operator==(const Options& value) const
{
	//stack
	if (stackProfileEnabled != value.stackProfileEnabled) return false;
	if (stackResolve != value.stackResolve) return false;
	if (stackMode != value.stackMode) return false;
	if (stackLibunwind != value.stackLibunwind) return false;
	if (stackSkip != value.stackSkip) return false;
	//time
	if (this->timeProfileEnabled != value.timeProfileEnabled) return false;
	if (this->timeProfilePoints != value.timeProfilePoints) return false;
	if (this->timeProfileLinear != value.timeProfileLinear) return false;
	//output
	if (this->outputName != value.outputName) return false;
	if (this->outputIndent != value.outputIndent) return false;
	if (this->outputJson != value.outputJson) return false;
	if (this->outputLua != value.outputLua) return false;
	if (this->outputCallgrind != value.outputCallgrind) return false;
	if (this->outputDumpConfig != value.outputDumpConfig) return false;
	if (this->outputVerbosity != value.outputVerbosity)  return false;
	if (this->outputStackTree != value.outputStackTree) return false;
	if (this->outputLoopSuppress != value.outputLoopSuppress) return false;
	//max stack
	if (this->maxStackEnabled != value.maxStackEnabled) return false;
	//maps
	if (this->distrAllocSize != value.distrAllocSize) return false;
	if (this->distrReallocJump != value.distrReallocJump) return false;
	//trace
	if (this->traceEnabled != value.traceEnabled) return false;
	//info
	if (this->infoHidden != value.infoHidden) return false;
	//exe
	if (this->exe != value.exe) return false;
	if (this->childs != value.childs) return false;
	if (this->enabled != value.enabled) return false;
	//dump
	if (this->dumpOnSignal != value.dumpOnSignal) return false;
	if (this->dumpAfterSeconds != value.dumpAfterSeconds) return false;
	
	return true;
}

/**********************************************************/
/**
 * Load values from string, mostly to be used from MALT_OPTION environment variable.
 * 
 * It expect string format like :
 * 
 * SEC1:NAME1=VALUE1;SEC2:NAME2=VALUE2;
 * 
 * @param value Define the string to load as a config file.
**/
void Options::loadFromString ( const char* value )
{
	//trivial
	if (value == NULL)
		return;

	//create fake dictionary
	dictionary * dic = dictionary_new(10);
	
	//copy string
	char * dump = strdup(value);
	
	//loop on separators ';'
	char * cur = dump;
	while (*cur != '\0')
	{
		//remind start
		char * start = cur;
		char * sep = NULL;
		
		//search ';' or '\0'
		while (*cur != ';' && *cur != '\0')
		{
			if (*cur == '=')
				sep = cur;
			cur++;
		}
		
		//skip to next
		if (cur == start)
		{
			cur++;
			continue;
		}
		
		//is end
		bool isEnd = (*cur == '\0');
		assumeArg(sep != NULL,"Invalid string format to setup option : '%1', expect SECTION:NAME=VALUE.").arg(start).end();
		
		//cut strings
		*cur = '\0';
		*sep = '\0';
		sep++;
		
		//setup in INI
		IniParserHelper::setEntry(dic,start,sep);
		
		//move
		if (isEnd == false)
			cur++;
	}
	
	//load
	this->loadFromIniDic(dic);

	//free
	iniparser_freedict(dic);
	free(dump);
}

/**********************************************************/
/**
 * Internal function to load options from iniDic.
**/
void Options::loadFromIniDic ( dictionary* iniDic )
{
	//errors
	assert(iniDic != NULL);
	
	//load values for time profiling
	this->timeProfileEnabled  = iniparser_getboolean(iniDic,"time:enabled",this->timeProfileEnabled);
	this->timeProfilePoints   = iniparser_getint(iniDic,"time:points",this->timeProfilePoints);
	this->timeProfileLinear   = iniparser_getboolean(iniDic,"time:linear-index",this->timeProfileLinear);
	
	//load values for stack profiling
	this->stackResolve        = iniparser_getboolean(iniDic,"stack:resolve",this->stackResolve);
	this->stackProfileEnabled = iniparser_getboolean(iniDic,"stack:enabled",this->stackProfileEnabled);
	this->stackLibunwind      = iniparser_getboolean(iniDic,"stack:libunwind",this->stackLibunwind);
	this->stackMode           = iniparser_getstring(iniDic,"stack:mode",(char*)this->stackMode.c_str());
	this->stackSkip           = iniparser_getint(iniDic,"stack:skip",this->stackSkip);
	
	//load values for output
	this->outputName          = iniparser_getstring(iniDic,"output:name",(char*)this->outputName.c_str());
	this->outputIndent        = iniparser_getboolean(iniDic,"output:indent",this->outputIndent);
	this->outputJson          = iniparser_getboolean(iniDic,"output:json",this->outputJson);
	this->outputLua           = iniparser_getboolean(iniDic,"output:lua",this->outputLua);
	this->outputCallgrind     = iniparser_getboolean(iniDic,"output:callgrind",this->outputCallgrind);
	this->outputDumpConfig    = iniparser_getboolean(iniDic,"output:config",this->outputDumpConfig);
	this->outputVerbosity     = iniparser_getverbosity(iniDic,"output:verbosity",this->outputVerbosity);
	this->outputStackTree     = iniparser_getboolean(iniDic,"output:stack-tree",this->outputStackTree);
	this->outputLoopSuppress  = iniparser_getboolean(iniDic,"output:loop-suppress",this->outputLoopSuppress);
	
	//max stack
	this->maxStackEnabled     = iniparser_getboolean(iniDic,"max-stack:enabled",this->maxStackEnabled);
	
	//maps
	this->distrAllocSize      = iniparser_getboolean(iniDic,"distr:alloc-size",this->distrAllocSize);
	this->distrReallocJump    = iniparser_getboolean(iniDic,"distr:realloc-jump",this->distrReallocJump);
	
	//trace
	this->traceEnabled        = iniparser_getboolean(iniDic,"trace:enabled",this->traceEnabled);
	
	//info
	this->infoHidden          = iniparser_getboolean(iniDic,"info:hidden",this->infoHidden);

	//filter
	this->exe                 = iniparser_getstring(iniDic,"filter:exe",(char*)this->exe.c_str());
	this->childs              = iniparser_getboolean(iniDic,"filter:childs",this->childs);
	this->enabled             = iniparser_getboolean(iniDic,"filter:enabled",this->enabled);

	//dump
	this->dumpOnSignal        = iniparser_getstring(iniDic,"dump:on-signal",(char*)this->dumpOnSignal.c_str());
	this->dumpAfterSeconds    = iniparser_getint(iniDic,"dump:after-seconds",this->dumpAfterSeconds);
}

/**********************************************************/
/**
 * Function to load options from a config file in INI format.
**/
void Options::loadFromFile(const char* fname)
{
	//load ini dic
	dictionary * iniDic;
	assert(fname != NULL);
	iniDic = iniparser_load(fname);
	
	//errors
	assumeArg(iniDic != NULL,"Failed to load config file : %1 !").arg(fname);
	
	//load
	loadFromIniDic(iniDic);
	
	//free dic
	iniparser_freedict(iniDic);
	
	//TODO apply getenv MALT_OPTIONS to override here and add "envOverride" parameter to enable it from caller
}

/**********************************************************/
/**
 * Helper function to convert the options to JSON output format and dump it
 * into the MALT output profile.
**/
void convertToJson(htopml::JsonState & json,const Options & value)
{
	json.openStruct();
		json.openFieldStruct("time");
			json.printField("enabled",value.timeProfileEnabled);
			json.printField("points",value.timeProfilePoints);
			json.printField("linear",value.timeProfileLinear);
		json.closeFieldStruct("time");
		
		json.openFieldStruct("stack");
			json.printField("enabled",value.stackProfileEnabled);
			json.printField("mode",value.stackMode.c_str());
			json.printField("resolve",value.stackResolve);
			json.printField("libunwind",value.stackLibunwind);
			json.printField("stackSkip",value.stackSkip);
		json.closeFieldStruct("stack");
		
		json.openFieldStruct("output");
			json.printField("callgrind",value.outputCallgrind);
			json.printField("dumpConfig",value.outputDumpConfig);
			json.printField("index",value.outputIndent);
			json.printField("json",value.outputJson);
			json.printField("lua",value.outputLua);
			json.printField("name",value.outputName);
			json.printField("verbosity",verbosityToString(value.outputVerbosity));
			json.printField("stackTree",value.outputStackTree);
			json.printField("loopSuppress",value.outputLoopSuppress);
		json.closeFieldStruct("output");
		
		json.openFieldStruct("maxStack");
			json.printField("enabled",value.maxStackEnabled);
		json.closeFieldStruct("maxStack");
		
		json.openFieldStruct("distr");
			json.printField("allocSize",value.distrAllocSize);
			json.printField("reallocJump",value.distrReallocJump);
		json.closeFieldStruct("distr");
		
		json.openFieldStruct("info");
			json.printField("hidden",value.infoHidden);
		json.closeFieldStruct("info");

		json.openFieldStruct("filter");
			json.printField("exe",value.exe);
			json.printField("childs",value.childs);
			json.printField("enabled",value.enabled);
		json.closeFieldStruct("filter");

		json.openFieldStruct("dump");
			json.printField("onSignal", value.dumpOnSignal);
			json.printField("afterSeconds", value.dumpAfterSeconds);
		json.closeFieldStruct("dump");
	json.closeStruct();
}

/**********************************************************/
/**
 * Helper to dump the config as INI file.
**/
void Options::dumpConfig(const char* fname)
{
	//create dic
	assert(fname != NULL);
	dictionary * dic = dictionary_new(10);
	
	//fill
	IniParserHelper::setEntry(dic,"time:enabled",this->timeProfileEnabled);
	IniParserHelper::setEntry(dic,"time:points",this->timeProfilePoints);
	IniParserHelper::setEntry(dic,"time:linear-index",this->timeProfileLinear);
	
	//stack
	IniParserHelper::setEntry(dic,"stack:enabled",this->timeProfileEnabled);
	IniParserHelper::setEntry(dic,"stack:mode",this->stackMode.c_str());
	IniParserHelper::setEntry(dic,"stack:resolve",this->stackResolve);
	IniParserHelper::setEntry(dic,"stack:libunwind",this->stackLibunwind);
	IniParserHelper::setEntry(dic,"stack:skip",this->stackSkip);
	
	//output
	IniParserHelper::setEntry(dic,"output:name",this->outputName.c_str());
	IniParserHelper::setEntry(dic,"output:lua",this->outputLua);
	IniParserHelper::setEntry(dic,"output:json",this->outputJson);
	IniParserHelper::setEntry(dic,"output:callgrind",this->outputCallgrind);
	IniParserHelper::setEntry(dic,"output:indent",this->outputIndent);
	IniParserHelper::setEntry(dic,"output:config",this->outputDumpConfig);
	IniParserHelper::setEntry(dic,"output:verbosity",verbosityToString(this->outputVerbosity));
	IniParserHelper::setEntry(dic,"output:stack-tree",this->outputStackTree);
	IniParserHelper::setEntry(dic,"output:loop-suppress",this->outputLoopSuppress);
	
	//max stack
	IniParserHelper::setEntry(dic,"max-stack:enabled",this->maxStackEnabled);
	
	//maps
	IniParserHelper::setEntry(dic,"distr:alloc-size",this->distrAllocSize);
	IniParserHelper::setEntry(dic,"distr:realloc-jump",this->distrReallocJump);
	
	//trace
	IniParserHelper::setEntry(dic,"trace:enabled",this->traceEnabled);
	
	//info
	IniParserHelper::setEntry(dic,"info:hidden",this->infoHidden);

	//exe
	IniParserHelper::setEntry(dic,"filter:exe",this->exe.c_str());
	IniParserHelper::setEntry(dic,"filter:childs",this->childs);
	IniParserHelper::setEntry(dic,"filter:enabled",this->enabled);

	//dump
	IniParserHelper::setEntry(dic,"dump:on-signal",this->dumpOnSignal.c_str());
	IniParserHelper::setEntry(dic,"dump:after-seconds",this->dumpAfterSeconds);
	
	//write
	FILE * fp = fopen(fname,"w");
	assumeArg(fp != NULL,"Failed to write dump of config file into %1 : %2 !").arg(fname).argStrErrno().end();
	iniparser_dump_ini(dic,fp);
	fclose(fp);
	
	//free
	iniparser_freedict(dic);
}

/**********************************************************/
/**
 * Internal function to split strings on ':' and extract the section name.
**/
std::string IniParserHelper::extractSectionName ( const char * key )
{
	std::string tmp;
	int i = 0;
	while (key[i] != ':' && key[i] != '\0')
		tmp += key[i++];
	return tmp;
}

/**********************************************************/
/**
 * Updat some entries of a dictionnary.
 * @param dic Define the dictionnary to update.
 * @param key Define the key to update.
 * @param value Define the value to setup for the given key.
**/
void IniParserHelper::setEntry(dictionary* dic, const char* key, const char* value)
{
	iniparser_set(dic,extractSectionName(key).c_str(),NULL);
	iniparser_set(dic,key,value);
}

/**********************************************************/
/**
 * Help set setup ini dic entries from boolean by converting them to string
 * internally.
 * @param dic Define the dictionnary to fill.
 * @param key Define the key to update (key:name)
 * @param value Define the boolean value to setup.
**/
void IniParserHelper::setEntry(dictionary* dic, const char* key, bool value)
{
	setEntry(dic,key,value?"true":"false");
}

/**********************************************************/
/**
 * Help set setup ini dic entries from integer by converting them to string
 * internally.
 * @param dic Define the dictionnary to fill.
 * @param key Define the key to update (key:name)
 * @param value Define the integer value to setup.
**/
void IniParserHelper::setEntry(dictionary* dic, const char* key, int value)
{
	char buffer[64];
	sprintf(buffer,"%d",value);
	setEntry(dic,key,buffer);
}

/**********************************************************/
/**
 * Need to be call once after malloc is available.
**/
Options& initGlobalOptions ( void )
{
	//error
	assume (gblOptions == NULL,"initGlobalOptions was used previously, gblOptions is already init ! ");
	gblOptions = new Options();
	return *gblOptions;
}

/**********************************************************/
/**
 * Convert verbosity level from string.
**/
Verbosity verbosityFromString(const std::string & value)
{
	//search match
	for (int i = 0 ; i < MALT_VERBOSITY_COUNT ; i++)
		if (value == cstVerbosityLevels[i])
			return (Verbosity)i;

	//error
	MALT_FATAL_ARG("Invalid verbosity level: '%1'").arg(value).end();
	return MALT_VERBOSITY_DEFAULT;
}

/**********************************************************/
/**
 * Convert verbosity level to string.
**/
const char * verbosityToString(Verbosity value)
{
	assert(value < MALT_VERBOSITY_COUNT);
	return cstVerbosityLevels[value];
}

/**********************************************************/
Verbosity iniparser_getverbosity(dictionary * dic, const char * key, Verbosity notFound)
{
	//tansmit
	char * tmp = iniparser_getstring(dic,key,(char*)verbosityToString(notFound));

	//convert back
	Verbosity res = verbosityFromString(tmp);

	//ok
	return res;
}

/**********************************************************/
/**
 * Dump to stream to be used for json output.
**/
std::ostream & operator << (std::ostream & out, Verbosity value)
{
	out << verbosityToString(value);
	return out;
}

}
